File: doc/book/samples/nim/sample/src/sample.nim
import autd3
import autd3/soem

proc onLost(msg: cstring) =
    echo msg
    quit(-1)

when isMainModule:
    try:
        var geometry = initGeometryBuilder().addDevice([0.0, 0.0, 0.0], [0.0, 0.0, 0.0]).build()

        var soem = initSOEM()
        let link = soem.highPrecision(true).onLost(onLost).build()

        var autd = openController(geometry, link)

        autd.send(clear(), 20*1000*1000)
        autd.send(synchronize(), 20*1000*1000)

        let firmList = autd.firmwareInfoList()
        for firm in firmList:
            echo firm

        let config = initSilencerConfig()
        autd.send(config, 20*1000*1000)

        let center = autd.geometry.center
        let f = initFocus([center[0], center[1], 150.0])
        let m = initSine(150)

        autd.send(m, f, 20*1000*1000)

        discard stdin.readLine

        autd.close()

    except:
        let
            e = getCurrentException()
            msg = getCurrentExceptionMsg()
        echo "Got exception ", repr(e), " with message ", msg

File: doc/samples/julia/sample.jl
using Printf

using StaticArrays

using AUTD3

function on_lost(msg::Cstring)
    println(msg)
    exit(-1)
end

geometry = GeometryBuilder().add_device(SVector(0.0, 0.0, 0.0), SVector(0.0, 0.0, 0.0)).build()

link = SOEM().on_lost(on_lost).build()

cnt = Controller(geometry, link)

cnt.send(Clear())
cnt.send(Synchronize())

firm_info_list = cnt.firmware_info_list()
for firm_info in firm_info_list
    @printf("%s\n", firm_info)
end

const g = Focus(cnt.geometry().center() + SVector(0.0, 0.0, 150.0))
const m = Sine(150)

cnt.send(m, g)

readline()

cnt.close()